### PR TITLE
Add missing implementation for functions in Boogie AST

### PR DESCRIPTION
**Pull request by zhassan-aws**
Flesh-out functions in Boogie AST (both definition and writing).

Functions are needed for SMT builtins (will be added in following PR).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Review by celinval — COMMENTED** [review]
Do we currently emit a `Function`?

File: boogie_ast/src/boogie_program/mod.rs [hunk 1]
```diff
@@ -188,8 +194,27 @@ impl Procedure {
...
 /// A function in Boogie is a mathematical function (deterministic, has no side
 /// effects, and whose body is an expression)
-struct Function {}
+pub struct Function {
+    name: String,
+    parameters: Vec<Parameter>,
+    return_type: Type,
+    // a body is optional (e.g. SMT built-ins)
+    body: Option<Expr>,
+    // Boogie attributes, e.g. `{:bvbuiltin "bvnot}`
```
[the remark this answers is not in the document]
Comment (zhassan-aws) [reply boogie_ast/src/boogie_program/mod.rs]: Good catch! Fixed.

File: boogie_ast/src/boogie_program/writer.rs [hunk 2]
```diff
@@ -132,6 +132,41 @@ impl BoogieProgram {
...
+                write!(writer, ", ")?;
+            }
+            write!(writer, "{}: ", param.name)?;
+            param.typ.write_to(writer)?;
+        }
+        write!(writer, ") returns (")?;
+        self.return_type.write_to(writer)?;
+        write!(writer, ")")?;
+        if let Some(body) = &self.body {
+            writeln!(writer, " {{")?;
```
[the remark this answers is not in the document]
Comment (zhassan-aws) [reply boogie_ast/src/boogie_program/writer.rs:155]: This uses `writeln` which adds a newline (as opposed to `write`, which doesn't).

**Comment by zhassan-aws** [comment]
> Do we currently emit a `Function`?

Not yet. The next PR will use `Function`s to implement SMT bit-vector builtins.

**Pull request merged**